File: findKClosestElements.cpp
// Time Complexity : O (log n) + k --> log n for binary search and k to copy over the elements to the result 
// Space Complexity : O (1) 
// Did this code successfully run on Leetcode : Yes 

// Binary search for the lower bound or the start of the window 

// The window is of size k and the idea is to find the start of the window using binary search 
// low = 0
// high = length - k because we need k elements left from result to return 

// In binary search, we are comparing the mid with the target and moving low or high
// here the mid is the entire window 

// the window is from arr[mid] - arr[mid+k-1] but we consider an extra elemet 
// The extra element is to decide which of the boundary element is definitely not part of the result. If we had just considered k elements, there would be no way to determine which direction to move. If I have k+1 elemnts, I know if the first element is closer to x or the last and can accordingly move my low/high 

// low = mid + 1 -> when (arr[mid] - x) > (x - arr[mid+k])
// else 
// high = mid 


class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        if(arr.size() ==0) return arr;

        // Binary search initialise low, high

        int low = 0;
        int high = arr.size()-k;

        while(low<high){
            int mid = low + (high-low)/2; //to avoid integer overflow
            if(x-arr[mid] > arr[mid+k]-x)
                low = mid+1;
            else
                high = mid;
        }
        cout<<low<<high;
        vector<int> result (arr.begin()+low, arr.begin()+low+k);
        return result;
    }
};

File: Find k closest elements.py
# Time:- O(n)
# Space:- O(1)
# Approach:- Binary Search the number<=target and use two pointer and a deque to maintain sorted order. Compare left and right pointer values decrement the left
# pointer if the left pointer value is more closer to the target than the right append left the left value to a queue to maintain sorted order. Otherwise increment right and append right to the queue
from bisect import bisect_left
class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        def bs(left,right,x):
            while(left<right):
                mid=(left+right)//2
                if arr[mid]==x:
                    return mid
                if arr[mid]<x:
                    left=mid+1
                else:
                    right=mid-1
            return left-1
            
        left=bs(0,len(arr)-1,x)
        right=left+1
        res=collections.deque([])
        while(left>=0 and right<len(arr) and len(res)<k):
            if x-arr[left]<=arr[right]-x:
                # maitaining sorted order
                res.appendleft(arr[left])
                left-=1
            else:
                # maintaining sorted order
                res.append(arr[right])
                right+=1
        # we breached right boundary but we still have elements to put in the list
        while(left>=0 and len(res)<k):
            res.appendleft(arr[left])
            left-=1
        # we breached left boundary but we still have elements to put in the list
        while(right<len(arr) and len(res)<k):
            res.append(arr[right])
            right+=1
        return res